### PR TITLE
LUCENE-10520 / #11556 HTMLStripCharFilter bugfix

### PR DESCRIPTION
**Pull request by elliotzlin**
### Description
"<" and ">" characters are acceptable in attribute values per the HTML5 spec. See #11556 for more details.

### CODE REVIEW

**Comment by rmuir** [comment]
This change looks good to me! I just want to clone the branch and try to figure how you removed 25,000 lines of code from the generated jflex :)

**Comment by dweiss** [comment]
What?! How?! :) Seems like the automaton got a whole lot smaller:
```
   private static int [] zzUnpackAttribute() {
-    int [] result = new int[14794];
+    int [] result = new int[3082]; 
```

**Comment by elliotzlin** [comment]
I hadn't even noticed that! I hope it's not some JFlex side-effect that I unwittingly triggered with my change.

**Comment by rmuir** [comment]
I don't think you did anything wrong. The changes look fine, it is just a real puzzler.

I even cloned your repo and inspected the enormous diff to the generated code and I only see smaller DFAs, not any "logical changes". The test suite does pass.

**Comment by elliotzlin** [comment]
@rmuir @dweiss I wanted to follow up on this. Is there anything I should work on here or is this good to go?

**Comment by dweiss** [comment]
Sorry this slipped, @elliotzlin . I think it's fine. Please add a lucene/CHANGES.txt entry (it'll go towards 9.5.0). I'll handle the commit and backport.

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/charfilter/TestHTMLStripCharFilter.java [hunk 1]
```diff
@@ -632,6 +635,16 @@ public void testUTF16Surrogates() throws Exception {
     analyzer.close();
   }
 
+  public void testForIssue10520() throws IOException {
```
Comment (dweiss) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/charfilter/TestHTMLStripCharFilter.java:638]: Maybe add an explicit javadoc with a see tag pointing at the github issue here (full URL)?

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/charfilter/TestHTMLStripCharFilter.java [hunk 2]
```diff
@@ -632,6 +635,22 @@ public void testUTF16Surrogates() throws Exception {
     analyzer.close();
   }
 
+  /**
+   * Test that attributes with '>' or '<' characters are parsed correctly.
```
Comment (dweiss) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/charfilter/TestHTMLStripCharFilter.java:639]:
```suggestion
   * Test that attributes with {@code '>'} or {@code '<'} characters are parsed correctly.
```

**Comment by elliotzlin** [comment]
@dweiss can you help run the workflows again?

**Comment by elliotzlin** [comment]
@rmuir would you be able to help with running the workflows?

**Comment by elliotzlin** [comment]
@dweiss @rmuir apologies for reviving this ancient thread, but I was wondering if one of you can help run the PR checks again?

**Comment by dweiss** [comment]
I will take a look tomorrow - was out of office last week.

**Review by dweiss — APPROVED** [review]
LGTM. These broken HTML examples are so hairy that I found them hard to understand... if somebody wants to sanitize such HTML then a preflight with tagsoup-style html tidy-up tool would probably be a wise step.

**Comment by mjustice3** [comment]
I believe I've found a regression with this bugfix. A test case that exposes:

```
  public void testForIssue10520Regression() throws IOException {
    String test =
        "<!DOCTYPE html><html lang=\"en\"><head><title>Test</title></head><a href=\"https://www.somewhere.com?data=\">a link</a> some text <a href=\"https://www.elsewhere.com\">another link</a></html>";
    Reader reader = new StringReader(test);
    HTMLStripCharFilter filter = new HTMLStripCharFilter(reader);
    StringWriter result = new StringWriter();
    filter.transferTo(result);
    assertEquals("Test\n\na link some text another link", result.toString().trim());
  }
```

The problem is with the empty `data=` parameter at the end of the first url. We see a few of those in our document set and that is how I noticed.

**Pull request merged**